sota-launcher/src/datatypes.rs

```diff
@@ -0,0 +1,234 @@
+use std::str::FromStr;
+use toml;
+use uuid;
+
+
+error_chain!{
+    foreign_links {
+        Http(reqwest::Error);
+        Io(io::Error);
+        Toml(toml::de::Error);
+        Uuid(uuid::ParseError);
+    }
+
+    errors {
+        Config(s: String) {
+            description("config error")
+            display("config error: '{}'", s)
+        }
+    }
+}
+
+
+#[derive(PartialEq, Clone, Copy, Debug)]
+pub enum Environment {
+    CI,
+    QA,
+    Production,
+}
+
+impl FromStr for Environment {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self> {
+        match s.to_lowercase().as_ref() {
+            "ci" => Ok(Environment::CI),
+            "qa" => Ok(Environment::QA),
+            "production" => Ok(Environment::Production),
+            _ => Err(format!("unknown environment: {}", s).into())
+        }
+    }
+}
+
+impl Display for Environment {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match *self {
+            Environment::CI => write!(f, "https://ci-app.atsgarage.com/api/v1"),
+            Environment::QA => write!(f, "https://qa-app.atsgarage.com/api/v1"),
+            Environment::Production => write!(f, "https://app.atsgarage.com/api/v1"),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for Environment {
+    fn deserialize<D: Deserializer<'de>>(de: D) -> result::Result<Self, D::Error> {
+        if let json::Value::String(ref s) = Deserialize::deserialize(de)? {
+            s.parse().map_err(|err| serde::de::Error::custom(format!("{}", err)))
+        } else {
+            Err(serde::de::Error::custom("expected string type for environment"))
+        }
+    }
+}
+
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct UpdateTargets {
+    pub targets: HashMap<String, Update>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Update {
+    pub from: Option<UpdateTarget>,
+    pub to: UpdateTarget
+}
+
+#[allow(non_snake_case)]
+#[derive(Serialize, Deserialize, Debug)]
+pub struct UpdateTarget {
+    pub target: String,
+    pub targetLength: u64,
+    pub checksum: Checksum,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Checksum {
+    pub method: ChecksumMethod,
+    pub hash: String,
+}
+
+#[derive(PartialEq, Clone, Copy, Debug)]
+pub enum ChecksumMethod {
+    Sha256,
+    Sha512,
+}
+
+impl FromStr for ChecksumMethod {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self> {
+        match s.to_lowercase().as_ref() {
+            "sha256" => Ok(ChecksumMethod::Sha256),
+            "sha512" => Ok(ChecksumMethod::Sha512),
+            _ => Err(format!("unknown checksum method: {}", s).into())
+        }
+    }
+}
+
+impl Serialize for ChecksumMethod {
+    fn serialize<S: Serializer>(&self, ser: S) -> result::Result<S::Ok, S::Error> {
+        ser.serialize_str(match *self {
+            ChecksumMethod::Sha256 => "sha256",
+            ChecksumMethod::Sha512 => "sha512",
+        })
+    }
+}
+
+impl<'de> Deserialize<'de> for ChecksumMethod {
```

Comment:
This can be rewritten

```rust
fn deserialize<D: Deserializer<'de>>(de: D) -> result::Result<Self, D::Error> {
  let s: String = Deserialize::deserialize(de)?;
  s.parse().map_err(|err| serde::de::Error::custom(format!("{}", err)))
}
```

To get the same effect, and then you're not tying yourself to `serde_json`.

But also, you could use `#[serde(rename = "XXX")]` on the variants and not have to have the manual `Serialize`, `Deserialize`, `ToString`, and `FromString` implementations.